public/app/features/scenes/core/serialization.ts

```diff
@@ -0,0 +1,311 @@
+  //   : undefined;
+
+  if (node instanceof NestedScene) {
+    return {
+      $ref: node.state.key,
+      type: node.constructor.name,
+    };
+  }
+
+  if (isDataNode(node)) {
+    return {
+      $ref: node.state.key,
+    };
+  }
+
+  if (isLayoutNode(node) && node.toJSON) {
+    if (node.state.children) {
+      return {
+        ...serializeNode(node),
+        children: (node as SceneObject<SceneLayoutState>).state.children.map((child) => serializeLayout(child)),
+      };
+    }
+
+    return serializeNode(node);
+    // {
+    //   ...node.toJSON(),
+    //   key: node.state.key,
+    //   type: node.constructor.name,
+    //   inputParams: serializedInputParams,
+    // };
+  }
+}
+
+export function serializeNode(node: SceneObject) {
+  const serializedInputParams = isParametrizedState(node.state)
+    ? serializeInputParams(node.state.inputParams)
+    : undefined;
+
+  const json = node.toJSON ? node.toJSON() : {};
+
+  return {
+    ...json,
+    key: node.state.key,
+    inputParams: { ...serializedInputParams },
+    type: node.constructor.name,
+  };
+}
+
+// TODO: types for serialized Scene
+export function serializeScene(scene: Scene | NestedScene, isNested = false, hoistDataNodes = false): any {
+  const children = Array.from(flattenSceneNodes(scene.state.children), ([_, v]) => v);
+  const dataNodesIdxs: number[] = [];
+  const dataNodesMap: Map<string, SerializedNode<any>> = new Map();
+
+  const nestedScenesIdxs: number[] = [];
+
+  children.forEach((child, idx) => {
+    if (isDataNode(child)) {
+      dataNodesIdxs.push(idx);
+    } else if (child instanceof NestedScene) {
+      nestedScenesIdxs.push(idx);
+    }
+  });
+
+  const nestedScenes: Record<string, SerializedScene> = nestedScenesIdxs.reduce((acc, idx) => {
+    return {
+      ...acc,
+      [children[idx].state.key!]: (children[idx] as NestedScene).toJSON(),
+    };
+  }, {});
+
+  dataNodesIdxs.forEach((idx) => {
+    const child = children[idx];
+    if (child.toJSON) {
+      dataNodesMap.set(child.state.key!, serializeNode(child));
+    }
+  });
+
+  // To avoid data nodes duplication the nested scene inputs can be hoisted to the top level
+  if (hoistDataNodes) {
+    // iterate over nestedScenes and hoist data nodes to the top level
+    for (const [, nestedScene] of Object.entries(nestedScenes)) {
+      for (let idx = 0; idx < nestedScene.inputs.length; idx++) {
+        const nestedDataNode = nestedScene.inputs[idx];
+        const dataNode = dataNodesMap.get(nestedDataNode.key);
+        if (dataNode) {
+          if (nestedDataNode && isEqual(dataNode, nestedDataNode)) {
+            nestedScene.inputs[idx] = { $ref: dataNode.key };
+          }
+        }
+      }
+    }
+  }
+
+  return {
+    key: scene.state.key,
+    title: scene.state.title,
+    data: 'TODO: serialize Scene data?',
+    inputs: Array.from(dataNodesMap.values()),
+    layout: {
+      root: serializeLayout(scene, isNested).layout,
+      ...nestedScenes,
+    },
```

Comment:
should the nested scene not be serialized like any node?

Reply:
It should and it is (with the unfortunate `actions` property being serialized additionally - that's why I don't like it and think it will require a change!). What I want to do here is to have "reusable nested scenes". Self contained sub-scenes that one doesn't have to author somewhere deep in the root layout structure, but instead reference as a composable building blocks.